yt_dlp/extractor/pialive.py

```diff
@@ -0,0 +1,97 @@
+    PIA_LIVE_API_URL = 'https://api.pia-live.jp'
+    _VALID_URL = r'https?://player\.pia-live\.jp/stream/(?P<id>[\w-]+)'
+
+    _TESTS = [
+        {
+            'url': 'https://player.pia-live.jp/stream/4JagFBEIM14s_hK9aXHKf3k3F3bY5eoHFQxu68TC6krUDqGOwN4d61dCWQYOd6CTxl4hjya9dsfEZGsM4uGOUdax60lEI4twsXGXf7crmz8Gk__GhupTrWxA7RFRVt76',
+            'info_dict': {
+                'id': '88f3109a-f503-4d0f-a9f7-9f39ac745d84',
+                'display_id': '2431867_001',
+                'title': 'こながめでたい日２０２４の視聴ページ | PIA LIVE STREAM(ぴあライブストリーム)',
+                'live_status': 'was_live',
+                'comment_count': int,
+            },
+            'params': {
+                'getcomments': True,
+                'skip_download': True,
+                'ignore_no_formats_error': True,
+            },
+        },
+        {
+            'url': 'https://player.pia-live.jp/stream/4JagFBEIM14s_hK9aXHKf3k3F3bY5eoHFQxu68TC6krJdu0GVBVbVy01IwpJ6J3qBEm3d9TCTt1d0eWpsZGj7DrOjVOmS7GAWGwyscMgiThopJvzgWC4H5b-7XQjAfRZ',
+            'info_dict': {
+                'id': '9ce8b8ba-f6d1-4d1f-83a0-18c3148ded93',
+                'display_id': '2431867_002',
+                'title': 'こながめでたい日２０２４の視聴ページ | PIA LIVE STREAM(ぴあライブストリーム)',
+                'live_status': 'was_live',
+                'comment_count': int,
+            },
+            'params': {
+                'getcomments': True,
+                'skip_download': True,
+                'ignore_no_formats_error': True,
+            },
+        },
+    ]
+
+    def _extract_vars(self, variable, html):
+        return self._search_regex(
+            rf'(?:var|const)\s+{variable}\s*=\s*(["\'])(?P<value>(?:(?!\1).)+)\1',
+            html, 'variable', group='value', default=None)
+
+    def _real_extract(self, url):
+        video_key = self._match_id(url)
+        webpage = self._download_webpage(url, video_key)
+
+        program_code = self._extract_vars('programCode', webpage)
+        article_code = self._extract_vars('articleCode', webpage)
+
+        prod_configure = self._download_webpage(
+            self.PLAYER_ROOT_URL + self._search_regex(
+                r'<script[^>]+src=(["\'])(?P<url>/statics/js/s_prod\?(?:(?!\1).)+)\1',
+                webpage, 'prod configure page url', group='url'),
+            program_code, headers={'Referer': self.PLAYER_ROOT_URL},
+            note='Fetching prod configure page', errnote='Unable to fetch prod configure page')
+
+        payload, content_type = multipart_encode({
+            'play_url': video_key,
+            'api_key': self._extract_vars('APIKEY', prod_configure)})
+
+        player_tag_list = self._download_json(
+            f'{self.PIA_LIVE_API_URL}/perf/player-tag-list/{program_code}', program_code,
+            data=payload, headers={'Content-Type': content_type, 'Referer': self.PLAYER_ROOT_URL},
+            note='Fetching player tag list', errnote='Unable to fetch player tag list')
```

Comment:
I don't understand why a warning needs to be issued for 404s here. The player URLs are all obtained from this JSON, and if they cannot be retrieved, the request would fail anyway. And it seems that this url doesn't require login for the time being.

Reply:
> ```python
> webpage = self._download_webpage(url, video_key)
> 
> ```

Will we encounter a 404 here when extracting an expired URL?

Reply:
> > ```python
> > webpage = self._download_webpage(url, video_key)
> > ```
> 
> Will we encounter a 404 here when extracting an expired URL?

No, the player page is still accessible for now. It will become 404 in the future.

<br>

Since the webpage is rendered by the server, we can check if an event is expired by extracting the information from it:

```HTML
<div class="mainView">
  <section class="player comment-none">
    <div class="player-section">
      <div id="msg-player" class="msg-player error-message display-none">
        <p class="error"></p>
        <p class="help display-none" wovn-wait="">運営者へのお問い合わせ先は、<a href="https://pia-live.jp/help?lang=en" target="_blank">ヘルプ</a> よりご確認ください。</p>
      </div>
      <div class="player-wrapper">
        <div class="player-width">
                    <div class="play-end" style="background-image: url('/1/image/player-end.png');"></div>
                  </div>
      </div>

```

An archive [^1] shows that there is no `play-end` div before expiration:

[^1]: https://web.archive.org/web/20240828123119/https://player.pia-live.jp/stream/JwkIiPuCHnoe6315eaNYZYLkJnXfasSUmlqzqiZGs9lhpi_OZqm1NZCB9CaxAWtKzv1FKpC4qS_dJrVSlpsrc3Ose2Kvwnenbkx0BvtKnNIaMdBnv-P6sBFL3ge1G_kq

```HTML
<div class="mainView">
  ...
      <div class="player-wrapper">
        <div class="player-width">
                    <div id="player"></div>
                  ...

```

So:

```Python
if get_element_html_by_class('play-end', webpage):
    # The title is the only thing we can extract. it's meaning less, isn't it?
    raise ExtractorError('The video is no longer available', expected=True)

```

.

Reply:
By checking an archive [^1], I think that we can also get the start timestamp of an upcoming event.

[^1]: https://web.archive.org/web/20240628164326/https://player.pia-live.jp/stream/kCIDF3oRMnuu-YcT1wrEpXJtao548aC9BcNL3BWtRg5YW7G8HRdcWkxXxxLh6sll5kjqAD1QBjW13eMDV_45v-PuZnn1lrPzPvqrXxIwnVIqknUA-HFeKbl8Py4P628l

```HTML
<div class="mainView">
  ...
      <div class="player-wrapper">
        <div class="player-width">
                    <div class="play-waiting" style="background-image: url(https://image.pia-live.jp/110);">
            <div class="play-waiting__date">
              <span class="not-apply" wovn-ignore>2024/7/7</span><span>(日)</span><span class="not-apply" wovn-ignore>16:30</span><br/>
              <span>開場予定</span>
            </div>
            <p>※時間になりましたら、一度画面をリロードしてください</p>
          </div>
                  </div>
      </div>

```

```Python
start_info = clean_html(get_element_by_class('play-waiting__date', webpage))
date, time = self._search_regex(
    r'(?P<date>\d{4}/\d{1,2}/\d{1,2})\([月火水木金土日]\)(?P<time>\d{2}:\d{2})',
    start_info, 'start_info', fatal=False, group=('date', 'time'))
start_timestamp = unified_timestamp(f'{date} {time} +09:00')

```

Run the above code if we have no player URL found.

Reply:
Above checks have been implemented in a75a02ad2a96423f3bf87ba7310b3603bf00cb85.